src/txn/txn_ckpt.c

```diff
@@ -658,10 +661,34 @@ __checkpoint_prepare(WT_SESSION_IMPL *session, bool *trackingp, const char *cfg[
     }
 
     /*
-     * Refresh our snapshot here without publishing our shared ids to the world, doing so prevents
-     * us from racing with the stable timestamp moving ahead of current snapshot. i.e. if the stable
-     * timestamp moves after we begin the checkpoint transaction but before we set the checkpoint
-     * timestamp we can end up missing updates in our checkpoint.
+     * Wait for any transactions that are supposed to be stable to finish committing. This prevents
+     * a race where a transaction can begin committing at a time past stable, and another thread
+     * moves stable past that transaction's commit time, and we start checkpointing before it
+     * finishes; we need it in this checkpoint, but it won't be unless we wait for it. If we do,
+     * then this race reduces to the race described in the next comment.
+     *
+     * Note that arguably the proper solution for this race is to not allow stable to advance past a
+     * transaction that hasn't finished committing (or to make it wait instead of us) but that is
+     * not currently feasible.
+     */
+    if (use_timestamp && txn_global->meta_ckpt_timestamp != WT_TS_NONE) {
+        while (__wt_txn_checkpoint_cannot_start(session)) {
+            __wt_sleep(0, 100 * WT_THOUSAND);
+            WT_STAT_CONN_INCRV(session, txn_checkpoint_prep_wait, 100);
+            wait_time++;
+        }
+        /* Grumble (with timing data) if we had to wait more than five seconds. */
+        if (wait_time > 50)
+            __checkpoint_verbose_track(
+              session, "Finished waiting for necessary transactions to commit");
```

Comment:
This message indicates that we successfully waited until all the necessary transactions to commit, but actually we didn't. How about `Skip ...`?

Reply:
We did, or should have -- that check is only a print, not a timeout. However, maybe there should be a timeout here...